netsim/extra/evpn.multihoming/plugin.py

```diff
@@ -0,0 +1,206 @@
+            # update
+            _esi_stats.esi_lacp_id_info[esi_name][lacp_sys_id].append(f"{node.name}->{interface.ifname}")
+        else:
+            # create
+            _esi_stats.esi_lacp_id_info[esi_name][lacp_sys_id] = [ f"{node.name}->{interface.ifname}" ]
+    return
+
+def _generate_esi_warnings() -> None:
+    global _esi_stats
+    # ESI used only once
+    for esi_name, esi_use in _esi_stats.esi_reference.items():
+        if len(esi_use) == 1:
+            log.warning(
+                text=f'EVPN Ethernet Segment "{esi_name}" seems to be used only once ({esi_use[0]})',
+                module=_config_name,
+                flag='evpn_mh_segment_used_once',
+                )
+    # ESI LAG LACP mismatch
+    for esi_name, esi_lacp_ids in _esi_stats.esi_lacp_id.items():
+        if len(esi_lacp_ids) != 1:
+            mismatch_list = []
+            for m in list(esi_lacp_ids):
+                found_interfaces=",".join(_esi_stats.esi_lacp_id_info[esi_name][m])
+                mismatch_list.append("[{}]->[{}]".format(m, found_interfaces))
+            mismatch_txt = ", ".join(mismatch_list)
+            log.warning(
+                text=f'EVPN ESI-LAG "{esi_name}" LACP mismatch. Found: {mismatch_txt}',
+                module=_config_name,
+                flag='evpn_mh_lacp_mismatch',
+                )
+    return
+
+def post_transform(topology: Box) -> None:
+    global _config_name
+    global _auto_segments
+    global _esi_stats
+    _init_es_id_generator(topology)
+    for node in topology.nodes.values():
+        if not 'evpn' in node.module: continue
+        features = devices.get_device_features(node,topology.defaults)
+        es_supported = 'evpn.multihoming' in features
+        if not es_supported: continue
+        # Load Ethernet Segments data and expand it with auto segments
+        es_data = node.get('evpn.ethernet_segments', {})
+        es_data.update(_auto_segments)
+        for intf in node.get('interfaces',[]):
+            intf_es = intf.get('evpn.es', '')
+            if not intf_es: continue
+            # check for unreferenced ESI, and auto generate values
+            if intf_es not in es_data:
+                intf_es_data = _generate_es_struct(intf_es)
+                if log.debug_active('plugin'):
+                    print(f'evpn.multihoming: Found interface ESI definition without reference: '
+                          f'{intf_es} ({node.name}->{intf.ifname}): creating structure. [{intf_es_data}]')
+                es_data.update({ intf_es: intf_es_data })
+            # If LAG, auto generated ESI-ID and LACP-System-ID are not present, use also auto generated LACP ID
+            if intf.type == 'lag' and not intf.get('lag.lacp_system_id',False) and '_lacp_system_id' in es_data[intf_es]:
+                intf.lag.lacp_system_id = es_data[intf_es]._lacp_system_id
+            # Error Validation checks
+            if intf_es and intf.type not in _es_supported_on:
```

Comment:
Again, for the future: I would put the validation checks into a separate function. This thing is becoming hard to read, particularly sorting the "what's really going on" part from the validation part.